### PR TITLE
Requirements pertaining to storage

### PR DESCRIPTION
**Pull request by csarven**
In response to the following (and possibly other) issues on the relationship between data pod and Storage, determining and discovering storage:
* https://github.com/solid/data-interoperability-panel/issues/10
* https://github.com/solid/specification/issues/153
* https://github.com/solid/specification/issues/197

Special shoutout to @megoth and co sharing their implementation experience as well as documenting their findings: https://megoth.inrupt.net/public/SolidDataBrowser/pod-info.html ( https://web.archive.org/web/20201106145438/https://megoth.inrupt.net/public/SolidDataBrowser/pod-info.html ). More of this please! Thanks.

### CODE REVIEW

**Review by acoburn — APPROVED** [review]
This is a very useful clarification

**Review by dmitrizagidulin — APPROVED** [review]
👍 
This is great.

File: main/resource-access.bs [hunk 1]
```diff
@@ -95,6 +95,44 @@ Behaviour pertaining to authorization MUST proceed this optional redirect
 
 ## Linked Data ## {#linked-data}
 
+### Storage ### {#storage}
+
+When a servers supports a data pod, it MUST provide one or more storages
```
Comment (RubenVerborgh) [inline main/resource-access.bs:100]:
```suggestion
When a server supports a data pod, it MUST provide one or more storages
```

File: main/resource-access.bs [hunk 2]
```diff
@@ -95,6 +95,44 @@ Behaviour pertaining to authorization MUST proceed this optional redirect
 
 ## Linked Data ## {#linked-data}
 
+### Storage ### {#storage}
```
Comment (RubenVerborgh) [inline main/resource-access.bs:98]: Note: I was initially confused by the term, thinking that it referred to backends. Perhaps a small clarification about what it is and is not. The definition "a storage is the root container for all of its contained" is technically correct, but does not inform people who have never seen `pim:Storage`.

File: main/resource-access.bs [hunk 3]
```diff
@@ -95,6 +95,44 @@ Behaviour pertaining to authorization MUST proceed this optional redirect
...
+When a servers supports a data pod, it MUST provide one or more storages
+(`pim:Storage`). A storage is the root container for all of its contained
+resources (see [[#resource-containment]]).
+
+When a server supports multiple storages, the URIs MUST be allocated to
+non-overlapping space.
+
+Servers exposing the storage resource MUST advertise by including the HTTP
+`Link` header with `rel="type"` targeting
+`http://www.w3.org/ns/pim/space#Storage` when responding to request URI.
```
Comment (RubenVerborgh) [inline main/resource-access.bs:109]: What request URI? All? The storage? Root?

File: main/resource-access.bs [hunk 4]
```diff
@@ -95,6 +95,44 @@ Behaviour pertaining to authorization MUST proceed this optional redirect
...
+
+Clients can determine a resource is of type storage by making an HTTP `HEAD`
+or `GET` request on the target URL, and check for the `Link` header with
```
Comment (RubenVerborgh) [inline main/resource-access.bs:112]:
```suggestion
or `GET` request on the target URL, and checking for the `Link` header with
```

File: main/resource-access.bs [hunk 5]
```diff
@@ -95,6 +95,44 @@ Behaviour pertaining to authorization MUST proceed this optional redirect
...
+`rel="type"` targeting `http://www.w3.org/ns/pim/space#Storage`.
+
+Clients can determine the storage of a resource by moving up the URI path
+hierarchy until the response includes a `Link` header with `rel="type"`
```
Comment (RubenVerborgh) [inline main/resource-access.bs:116]: Do we have a ref for "moving up hierarchy"? We probably have it documented somewhere else already.

File: main/resource-access.bs [hunk 6]
```diff
@@ -95,6 +95,44 @@ Behaviour pertaining to authorization MUST proceed this optional redirect
...
+targeting `http://www.w3.org/ns/pim/space#Storage`. Clients may check the root
+path of a URI for the storage claim at any time.
+
+Clients can discover a storage by making an HTTP `GET` request on the target
+URL to retrieve an RDF representation [[!RDF11-CONCEPTS]], whose encoded RDF
+graph contains a relation of type `http://www.w3.org/ns/pim/space#storage`.
+The object of the relation is the storage (`pim:Storage`).
```
Comment (RubenVerborgh) [inline main/resource-access.bs:123]: What does the subject mean?

File: main/resource-access.bs [hunk 7]
```diff
@@ -95,6 +95,44 @@ Behaviour pertaining to authorization MUST proceed this optional redirect
...
+
+[[Source](https://github.com/solid/data-interoperability-panel/issues/10#issuecomment-598694029)]
+[[Source](https://github.com/solid/specification/issues/153#issuecomment-624630022)]
+
+When using Web Access Control ([[#wac]]):
+
+The root container (`pim:Storage`) MUST have an ACL auxiliary resource
+associated to it. The associated ACL document MUST include an authorization
```
Comment (RubenVerborgh) [inline main/resource-access.bs:131]:
```suggestion
directly associated to it. The associated ACL document MUST include an authorization
```
?

**Review by RubenVerborgh — APPROVED** [review]
All good, just suggestions 👍

File: main/resource-access.bs (hunk 6 above)
Comment (csarven) [reply main/resource-access.bs:123]: Leaving it alone for now.. pim-space doesn't define a domain. Current practice is a profile's storage but we did use it for workspaces as well.

File: main/resource-access.bs (hunk 5 above)
Comment (csarven) [reply main/resource-access.bs:116]: I was trying to keep it minimal and close to 3986 as opposed to "containment" but that can also work. We have used "traverse" in discussions eg. "traverse up in the containment hierarchy". And you've used "construct the URL of the parent container by stripping it to the closest slash" in the issue of ACL inheritance algorithm for clients. A bit too verbose for my taste. I'm okay to define the behaviour but perhaps we should do it in one go once the other usage is presented?

File: main/resource-access.bs (hunk 3 above)
Comment (csarven) [reply main/resource-access.bs]: The storage. https://github.com/solid/specification/pull/208/commits/4356daf06a0ca2216b9cbf76a856e2c82badc6c5

File: main/resource-access.bs (hunk 5 above)
Comment (RubenVerborgh) [reply main/resource-access.bs:118]: Ah yes, I thought it already was.

File: main/resource-access.bs (hunk 2 above)
Comment (csarven) [reply main/resource-access.bs:98]: Updated with https://github.com/solid/specification/pull/208/commits/cd78a59126171ca27e593b51180353ae8044145f . May be a bit redundant but readable. Thanks.

**Comment by csarven** [comment]
Thanks Ruben!

Noting here that I've also discussed some of these requirements with @timbl and it was agreed - although not the exact wording in this PR. Happy to revise.

**Pull request merged**